conans/model/info.py

```diff
@@ -398,14 +404,31 @@ def default_std_matching(self):
         same as specifying None, packages are the same
         """
 
-        if self.full_settings.cppstd and \
-                self.full_settings.compiler and \
-                self.full_settings.compiler.version:
+        if self.full_settings.compiler and \
+            self.full_settings.compiler.version:
             default = cppstd_default(str(self.full_settings.compiler),
                                      str(self.full_settings.compiler.version))
-            if default == str(self.full_settings.cppstd):
+
+            if str(self.full_settings.cppstd) == default:
                 self.settings.cppstd = None
 
+            if str(self.full_settings.compiler.cppstd) == default:
+                def remove_cppstd(settings):
+                    try:
+                        settings.compiler.cppstd = None  # It's the default, assign None
+                    except AttributeError:
+                        # Settings can be different at the moment of executing this function
+                        pass
+                    finally:
+                        return settings
+                self._adjust_settings_for_std = remove_cppstd
+
     def default_std_non_matching(self):
         if self.full_settings.cppstd:
             self.settings.cppstd = self.full_settings.cppstd
+        self._adjust_settings_for_std = lambda u: u  # Do nothing
+
+    def _settings_sha(self):
+        settings = self.settings.copy()
```

Comment:
I suggest:
- Having a default adjust = None
- If adjust = None, no need to copy, no need to call it.
- The adjust would do the copy internally, and return the sha. In this way it is not possible to forget to copy, which would destroy the ``info`` original values.

Reply:
I agree with the `adjust=None`, but I prefer to leave the `copy` and `sha` outside the _adjust_ function, maybe we can add here more _adjustments_ only for computing the ID and this way we can create the pipeline (I'm not sure if it will make sense right now for the `build_settings` and `toolset` ones):

```python
    def _settings_sha(self):
        if self._adjust_settings_for_std:
            settings = self.settings.copy()
            settings = self._adjust_settings_for_std(settings)
            settings = self._adjust_other_thing(settings)
            settings = self._adjust_...(settings)
            settings = ....
            return settings.sha
        return self.settings.sha
```

Just tell me if you really think it is better to make the copy inside the function and return the sha 😉